pkg/gds/store/trtl/iterator.go

```diff
@@ -0,0 +1,222 @@
+	return i.iter.Prev()
+}
+
+func (i *iterWrapper) Error() error {
+	return i.iter.Error()
+}
+
+func (i *iterWrapper) Release() {
+	i.iter.Release()
+}
+
+type trtlIterator struct {
+	client        trtlpb.TrtlClient
+	values        []*trtlpb.KVPair
+	index         int
+	nextPageToken string
+	cursor        trtlpb.Trtl_CursorClient
+	cancel        context.CancelFunc
+	snapshot      bool
+	namespace     string
+	err           error
+}
+
+func NewTrtlIterator(client trtlpb.TrtlClient, snapshot bool, namespace string) (iter *trtlIterator) {
+	iter = &trtlIterator{
+		client:    client,
+		snapshot:  snapshot,
+		namespace: namespace,
+	}
+
+	var ctx context.Context
+	ctx, iter.cancel = getContext()
+
+	if snapshot {
+		request := &trtlpb.CursorRequest{
+			Namespace: namespace,
+		}
+		iter.cursor, iter.err = iter.client.Cursor(ctx, request)
+		iter.values = make([]*trtlpb.KVPair, 0)
+	} else {
+		defer iter.cancel()
+		request := &trtlpb.IterRequest{
+			Namespace: namespace,
+		}
+
+		var reply *trtlpb.IterReply
+		var err error
+		if reply, err = iter.client.Iter(ctx, request); err != nil {
+			iter.err = err
+			return
+		}
+		iter.values = reply.Values
+		iter.nextPageToken = reply.NextPageToken
+	}
+	iter.index = -1
+
+	return iter
+}
+
+func (i *trtlIterator) nextCursor() bool {
+	if i.cursor == nil {
+		i.err = errors.New("nil cursor encountered on Next()")
+		return false
+	}
+
+	i.index++
+	if i.index >= len(i.values) {
+		// Fetch a new value from the cursor
+		var value *trtlpb.KVPair
+		var err error
+		if value, err = i.cursor.Recv(); err != nil {
+			if err != io.EOF {
+				i.err = err
+			}
+			i.cancel()
+			return false
+		}
+		i.values = append(i.values, value)
+	}
+
+	return true
+}
+
+func (i *trtlIterator) nextIter() bool {
+	i.index++
+	if i.index >= len(i.values) && i.nextPageToken != "" {
+		ctx, cancel := getContext()
+		defer cancel()
+
+		// Retrieve the next page from the server
+		request := &trtlpb.IterRequest{
+			Namespace: i.namespace,
+			Options: &trtlpb.Options{
+				PageToken: i.nextPageToken,
+			},
+		}
+		var reply *trtlpb.IterReply
+		var err error
+		if reply, err = i.client.Iter(ctx, request); err != nil {
+			i.err = err
+			return false
+		}
+
+		// Add the new values to the slice
+		i.values = append(i.values, reply.Values...)
+		i.nextPageToken = reply.NextPageToken
+	} else if i.index >= len(i.values) {
+		// No more values from the iterator
+		i.index = len(i.values) - 1
+		return false
+	}
+
+	return true
+}
+
+func (i *trtlIterator) Next() bool {
+	if i.err != nil {
+		return false
+	}
+	if i.snapshot {
+		return i.nextCursor()
+	}
+	return i.nextIter()
+}
+
+func (i *trtlIterator) Prev() bool {
+	i.index--
+	if i.index < 0 {
+		i.index = 0
+		return false
+	}
+	return true
+}
+
+func (i *trtlIterator) Seek(key []byte) bool {
+	for current := i.Value(); bytes.Compare(current, key) < 0; current = i.Value() {
+		if !i.Next() {
+			return false
+		}
+	}
+	return true
+}
```

Comment:
This is a pretty decent implementation of forward Seek, but there are a lot of hidden gotchas here. I propose that instead of seeking our array of values in memory, that we make Seek callable only once, before a call to Next and that we pass the seek argument into the CursorRequest. This offloads Seek to the `Trtl` database. Note that means Seek won't be available in `Iter` mode.